src/clj/puppetlabs/services/jruby/jruby_puppet_core.clj

```diff
@@ -61,117 +62,91 @@
+    (let [{:keys [ruby-load-path http-client-ssl-protocols
+                  http-client-cipher-suites
+                  http-client-connect-timeout-milliseconds
+                  http-client-idle-timeout-milliseconds
+                  use-legacy-auth-conf]} config]
+      (when-not ruby-load-path
+        (throw (Exception.
+                "JRuby service missing config value 'ruby-load-path'")))
+      (let [scripting-container (:scripting-container jruby-instance)
+            ruby-puppet-class (.runScriptlet scripting-container "Puppet::Server::Master")
+            puppet-config (jruby-internal/config->puppet-config config)
+            puppetserver-config (HashMap.)
+            env-registry (puppet-env/environment-registry)]
+        (when http-client-ssl-protocols
+          (.put puppetserver-config "ssl_protocols" (into-array String http-client-ssl-protocols)))
+        (when http-client-cipher-suites
+          (.put puppetserver-config "cipher_suites" (into-array String http-client-cipher-suites)))
+        (doto puppetserver-config
+          (.put "profiler" profiler)
+          (.put "environment_registry" env-registry)
+          (.put "http_connect_timeout_milliseconds" http-client-connect-timeout-milliseconds)
+          (.put "http_idle_timeout_milliseconds" http-client-idle-timeout-milliseconds)
+          (.put "use_legacy_auth_conf" use-legacy-auth-conf))
+        (let [jruby-puppet (.callMethodWithArgArray
+                            scripting-container
+                            ruby-puppet-class
+                            "new"
+                            (into-array Object
+                                        [puppet-config puppetserver-config])
+                            JRubyPuppet)]
+          (-> jruby-instance
+              (assoc :jruby-puppet jruby-puppet)
+              (assoc :environment-registry env-registry)))))))
+
+(schema/defn initialize-scripting-container-fn
+  [scripting-container
+   config :- jruby-schemas/JRubyConfig]
+  (doto scripting-container
+    (.setEnvironment (jruby-internal/managed-environment (jruby-internal/get-system-env) (:gem-home config)))
+    (.runScriptlet "require 'jar-dependencies'")
+    (.runScriptlet "require 'puppet/server/master'"))
+  scripting-container)
+
+(schema/defn cleanup-fn
+  [instance]
+  (.terminate (:jruby-puppet instance)))
+
+(schema/defn extract-jruby-config
+  [config :- {schema/Keyword schema/Any}]
+  (into {} (for [[key _] jruby-schemas/JRubyConfig]
+             {key (get config key)})))
 
-;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
-;;; Public
+(schema/defn extract-puppet-config
+  [config :- {schema/Keyword schema/Any}]
+  (into {} (for [[key _] jruby-puppet-schemas/JRubyPuppetConfig]
+             {key (get config key)})))
 
 (schema/defn ^:always-validate
-  initialize-config :- jruby-schemas/JRubyPuppetConfig
+  initialize-puppet-config :- jruby-puppet-schemas/JRubyPuppetConfig
```

Comment:
Right now the `JRubyPuppetConfig` schema has a bunch of overlap with the `JRubyConfig` in jruby-utils. In my opinion, at least, it would be better for this schema to only have the puppetserver-specific settings, and for this function to only initialize defaults for those settings. That might mean that we also want to do the extracting of jruby options and puppet options before we call this, so that from the start we have two separate maps? Not entirely sure about that, but kinda leaning towards the idea that that would make this a bit cleaner.